jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py

```diff
@@ -109,7 +108,11 @@ def __init__(
 
             self.num_dim = self._prefetch_data[0].shape[0]
             self.dtype = self._prefetch_data[0].dtype
-            self.index = self._build_index(vecs_iter)
+
+            if self.trained_index_file and not os.path.exists(self.trained_index_file):
+                raise ValueError(f'The trained index file {self.trained_index_file} does not exist')
+
+            self._build_index(vecs_iter)
```

Comment:
That might be a stupid question due to my lack of understanding Faiss, but: Why do we need to build the index here? Shouldn't that be already part of the training? I would expect, that when I run the training method and restart the Searcher, it can directly pick up the trained AND filled index from disc. Or is this not possible/performant?

Reply:
> Why do we need to build the index here? Shouldn't that be already part of the training? 

1) with providing the `trained_index_file`, we can load indexer from `trained_index_file`, and then start to build the indexer.
2) without providing the `trained_index_file`, there are two options:
    2.1  train the faiss indexer, and then use method 1) to build the indexer;
    2.2 start the query flow, the searcher will train the indexer by using the indexing data (loading from dump datafile). 

> I would expect, that when I run the training method and restart the Searcher, it can directly pick up the trained AND filled index from disc. 

Yes,  the current PR is actually implementing what you described above.

Reply:
I am pretty sure it isn't. The `_build_index` is still called on every startup. So let me rephrase this question on a different level:

Imagine you have a trained faiss model without data. And you have the same model filled with data and dump the filled index to disc (model + data is dumped).

Are the following two operations equally fast:

A. Loading the faiss model and populate it with data from a dump
B. just load the model + filled index from disc. No need to populate it with data, since it is already populated.

Is the B even possible?

If both are equally fast, we might still want to do B to not stress the PSQL database when to many shards ramp up.

Furthermore, doing B + incremental update would allow us to do performant query while indexing with short index-query delay.

Reply:
ah,  you are right. I got misunderstanding your question. Yes, doing B is a great feature, and is planned to approach it in the next PR.